### ISSUE TITLE
Timezone format in GPX files

### ISSUE THREAD

**Issue by teran** [merlos/iOS-Open-GPX-Tracker#132]
Hello,

First of all thank you for your app, it's awesome and helps me to set GPS coordinates to photos.

But when I tried to parse GPX files exported from the app with Go I've noticed your are using format very close to RFC3339/ISO8601 but 12-hours formatted with AM/PM markers which are not allowed normally (in RFC3339/ISO8601). Is there a reason for that behaviour or it should be RFC3339/ISO8601 ?

Thanks in advance.

**Comment by teran** [comment]
Btw, just tried to run [exiftool](https://www.sno.phy.queensu.ca/~phil/exiftool/) with GPX file produced by GPX Tracker and it's failed with `Warning: No track points found in GPS file`.
But if I manually convert the timestamps to 24-hours format right in XML - it works perfect.

**Comment by vincentneo** [comment]
Hi @teran 

This issue sounds concerning. I am in charge of the code that handles the GPX creation and stuff, so I think it is something that is part of that. Taking a look at one of my previously generated GPX file shows the time attribute marked as `2019-10-06T14:51:30Z` in one example, which looked like 24hr formatted to me. 

I don't know ISO8601 standard well enough to judge if that is right though, so please advice, and if possible, an example of how the time is stored in your file, would be great.

Just in case this is a version specific bug, may I also ask which version of the app are you using and iOS version?

**Comment by teran** [comment]
Hello @vincentneo, thank you for your reply!

The format you have is exactly rfc3339/ISO8601.
My gpx file I got via Share menu -> Save to dropbox is in attachment.

> Just in case this is a version specific bug, may I also ask which version of the app are you using and iOS version?
Sure, I'm running v1.6.0 from AppStore on iOS13.
And one more thing: I have 12-hours format in use on iOS for local time.

Thanks in advance.

[21-Sep-2019-357 PM.gpx.gz](https://github.com/merlos/iOS-Open-GPX-Tracker/files/3723526/21-Sep-2019-357.PM.gpx.gz)

**Comment by vincentneo** [comment]
Hi @teran

Very interesting bug indeed, this is nothing I’ve seen before. 

I’ve got a feeling that this has something to do with region. 
May I ask if the region set in your iPhone displays date and time different from other regions? 

I am asking so, as like you, I use 12hr time as well, have tested using App Store version, and using iOS 13, but was able to generate a valid formatted time.

Thank you.

**Comment by teran** [comment]
Hello @vincentneo

Not quite sure I got what mean on “displays date and time different from other regions?”. I’m using English localization and region set to Russia with 12-hour time format. ![here’s the screenshot](https://user-images.githubusercontent.com/1399459/66738729-d135e580-ee77-11e9-9927-cd08057150e0.png)

Please let me know if that’s what you asked about :)

**Comment by teran** [comment]
I've just made some experiments:

1) Switch to 24-hour format in iOS settings and export GPX file which was already collected: no results, still 12-hour format in XML
2) In addition to 24-hour format, switch localisation to Russian: the same result
3) Switch back to English (US) localisation but leave 24-hours format, collect new track and exported: works fine!

So it looks like 12-hours format in some conditions could affect internal timestamp format in the app, i.e. it's not about export it's about internal data storage.

**Comment by merlos** [comment]
![IMG_B8B6BC5EB84F-1](https://user-images.githubusercontent.com/404446/66746962-4d690280-ee51-11e9-8773-d827ead141b3.jpeg)

In my case time format is correct, but I use 24h time in the phone.

```
	<time>2019-09-10T14:16:15Z</time>
```
I went through different files and they´re good (as most of my gpx files were generated with dev versions).

Now, I have  #125 branch, I changed the settings and still getting things ok...

![image](https://user-images.githubusercontent.com/404446/66748560-74c1ce80-ee55-11e9-8014-53b34429eee0.png)


```
	<wpt lat="" lon="">
		<time>2019-10-14T13:32:36Z</time>
		<name>9:32:36 AM</name>
		<desc>Oct 14, 2019 at 9:32:36 AM</desc>
```

**Comment by teran** [comment]
@merlos you could try to reproduce exactly my settings: 12-hours format, language - English (US), region - Russia, calendar - Gregorian, temperature - Celsius.

Please note, it will work on new tracks only.

**Comment by merlos** [comment]
@teran. It did it. 

With your settings:

12-hours format, language - English (US), region - Russia, calendar - Gregorian, temperature - Celsius.

```
	<time>2019-10-14T2:48:14 PMZ</time>
		<name>10:48:14 AM</name>
		<desc>14 Oct 2019, 10:48:14 AM</desc>
	</wpt>
```
 

With your settings but 24h:
```
		<time>2019-10-14T14:52:28Z</time>
		<name>10:52:28</name>
		<desc>14 Oct 2019, 10:52:28</desc>
	</wpt>
```

**Comment by teran** [comment]
@merlos awesome! :) any ideas why this happens?

**Comment by merlos** [comment]
@teran,  I guess it is in the export method of CoreGPX the library that Open GPX tracker is using for import and export GPX files.

For example, for the Waypoint https://github.com/vincentneo/CoreGPX/blob/master/Classes/GPXWaypoint.swift, I guess is in `func addChildTag(toGPX gpx: NSMutableString, indentationLevel: Int)`

```
  self.addProperty(forValue: Convert.toString(from: time), gpx: gpx, tagName: "time", indentationLevel: indentationLevel) 
```
But @vincentneo is who coded it, for sure he will have a better idea of where to modify it.

**Comment by vincentneo** [comment]
@teran & @merlos,

I have fixed the issue in CoreGPX 0.7.1, tested to be working with ISO8601 formatted dates with Russian locale, 12hr time. :)

**Comment by teran** [comment]
@vincentneo awesome, thank you!

@merlos waiting for bugfix release?:)

**Comment by vincentneo** [comment]
@merlos, to make merging easier, I have added the fix to #125.

**Comment by teran** [comment]
Btw, in case of someone step into the issue and need to recover GPX files corrupted by the issue: I wrote a simple CLI tool to recover GPX files containing corrupted timestamps:

https://gist.github.com/teran/5464325ac3387a1851047093b03587f2

Could be run as:
```
go run ./recover-tool.go IN.gpx OUT.gpx
```

Since it's distributed as a source code Go compiler is required to run it.
